itests/fevm_test.go

```diff
@@ -48,22 +46,60 @@ func TestFEVMBasic(t *testing.T) {
 
 	idAddr, err := address.NewIDAddress(result.ActorID)
 	require.NoError(t, err)
-	t.Logf("actor ID address is %s", idAddr)
 
-	// invoke the contract with owner
-	{
-		entryPoint, err := hex.DecodeString("f8b2cb4f")
-		require.NoError(t, err)
+	return fromAddr, idAddr
+}
 
-		inputData, err := hex.DecodeString("000000000000000000000000ff00000000000000000000000000000000000064")
-		require.NoError(t, err)
+func invokeContract(t *testing.T, ctx context.Context, client *kit.TestFullNode, fromAddr address.Address, idAddr address.Address, funcSignature string, inputData []byte) []byte {
+	entryPoint := calcFuncSignature(funcSignature)
+	wait := client.EVM().InvokeSolidity(ctx, fromAddr, idAddr, entryPoint, inputData)
+	require.True(t, wait.Receipt.ExitCode.IsSuccess(), "contract execution failed")
+	result, err := cbg.ReadByteArray(bytes.NewBuffer(wait.Receipt.Return), uint64(len(wait.Receipt.Return)))
+	require.NoError(t, err)
+	return result
+}
 
-		wait := client.EVM().InvokeSolidity(ctx, fromAddr, idAddr, entryPoint, inputData)
+//function signatures are the first 4 bytes of the hash of the function name and types
+func calcFuncSignature(funcName string) []byte {
+	hasher := sha3.NewLegacyKeccak256()
+	hasher.Write([]byte(funcName))
+	hash := hasher.Sum(nil)
+	return hash[:4]
+}
 
-		require.True(t, wait.Receipt.ExitCode.IsSuccess(), "contract execution failed")
+//convert a simple byte array into input data which is a left padded 32 byte array
+func inputDataFromArray(input []byte) []byte {
+	inputData := make([]byte, 32)
+	copy(inputData[32-len(input):], input[:])
+	return inputData
+}
 
-		result, err := cbg.ReadByteArray(bytes.NewBuffer(wait.Receipt.Return), uint64(len(wait.Receipt.Return)))
-		require.NoError(t, err)
+//convert a "from" address into input data which is a left padded 32 byte array
+func inputDataFromFrom(t *testing.T, ctx context.Context, client *kit.TestFullNode, from address.Address) []byte {
+	fromId, err := client.StateLookupID(ctx, from, types.EmptyTSK)
+	require.NoError(t, err)
+
+	senderEthAddr, err := ethtypes.EthAddressFromFilecoinAddress(fromId)
```

Comment:
fromId is going to be an ID address and this API will always give you an 0xff eth address which isn't always accurate if we have an f4 address available

Reply:
sorry i am new don't understand this yet - should i make any changes?